README.md
#Plessc
Plessc is a gui frontend for lessc (or another compiler).  
This software require:  
* Python 3
* PyQT
* lessc (node-less on many distro or npm install lessc)
* All OS! ( *unix it's better)

##ScreenShot
![Screenshot](screenshot.png "1.1")

##Feature
* Log readable 
* Open the less file with the editor choosen
* Open all less file on the folder of the file choosen 
* Support minify and yui-compress of lessc
* Save a development version and a minimized (using the option)
* Auto check if file it's modified and compile
* Sourcemap support
* Lint in the log area

##Changelog

See the CHANGELOG.md file

##Run

Execute: plessc.py

CHANGELOG.md
#1.1
* [Feature] Sourcemap support
* [Feature] Lint in the log area
* [Feature] Autocomplete with history
* [Mini Feature] Wait cursor on compile
* [Bugfix] Correct file size showed
* [Bugfix] Log output
* [Removed] YUI Compress obsolete

#1.0 Beta
* Log readable 
* Open the less file with the editor choosen
* Open all less file on the folder of the file choosen 
* Support minify and yui-compress of lessc
* Save a development version and a minimized (using the option)
* Auto check if file it's modified and compile 


settings.py
#!/usr/bin/python3
from PyQt4.QtCore import *
from PyQt4.QtGui import *
import sys, os

from ui_settings import Ui_Settings

class SettingDialog ( QDialog , Ui_Settings):

	#var initialization
	settings = QSettings('Mte90','Plessc')
	settings.setFallbacksEnabled(False)

	def __init__ ( self, parent = None ):
		QDialog.__init__( self, parent )
		self.ui = Ui_Settings()
		self.ui.setupUi( self )
		self.ui.chooseLessc.clicked.connect(self.openLesscDialog)
		self.ui.chooseEditor.clicked.connect(self.openEditorDialog)
		self.ui.buttonBox.accepted.connect(self.saveSetting)
		#lessc path
		if self.settings.contains('lessc_path') == 'False' or len(self.settings.value('less_path')) == 0:
			self.settings.setValue('less_path','/usr/bin/lessc')
		#editor path
		if self.settings.contains('editor_path') == 'False' or len(self.settings.value('editor_path')) == 0:
			self.settings.setValue('editor_path','/usr/bin/kate')
		#checkbox for open all file in less file folder
		if self.settings.contains('less_folder') == 'False':
			self.settings.setValue('less_folder','False')
		if self.settings.value('less_folder') == 'False':
			self.ui.lessFolder.setChecked(False)
		else:
			self.ui.lessFolder.setChecked(True)

		self.ui.editor.setText(self.settings.value('editor_path'))
		self.ui.lesscPath.setText(self.settings.value('less_path'))
	
	#Function for open a dialog for choose the less compiler
	def openLesscDialog(self):
		lessc = QFileDialog.getOpenFileName(self, 'Choose less compiler',self.ui.lesscPath.text())
		self.ui.lesscPath.setText(self.lessc)
	
	#Function for open a dialog for choose the editor
	def openEditorDialog(self):
		editor = QFileDialog.getSaveFileName(self, 'Set editor',self.ui.editor.text())
		self.ui.editor.setText(editor)

	#Save the setting
	def saveSetting(self):
		if len(self.ui.editor.text()) > 0:
			self.settings.setValue('editor_path',self.ui.editor.text())
		else:
			self.ui.editor.setText('/usr/bin/kate')
			QMessageBox.critical(self.window(), "Editor empty","This field can't be empty!")
			
		if len(self.ui.lesscPath.text()) > 0:
			self.settings.setValue('less_path',self.ui.lesscPath.text())
		else:
			self.ui.lesscPath.setText('/usr/bin/lessc')
			QMessageBox.critical(self.window(), "Compiler empty","This field can't be empty!")
		
		if self.ui.lessFolder.isChecked() == False:
			self.settings.setValue('less_folder','False')
		else:
			self.settings.setValue('less_folder','True')

ui_settings.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file '/home/mte90/Desktop/Prog/Plessc/settings.ui'
#
# Created: Mon Aug 12 19:21:42 2013
#      by: PyQt4 UI code generator 4.10.2
#
# WARNING! All changes made in this file will be lost!

from PyQt4 import QtCore, QtGui

try:
    _fromUtf8 = QtCore.QString.fromUtf8
except AttributeError:
    def _fromUtf8(s):
        return s

try:
    _encoding = QtGui.QApplication.UnicodeUTF8
    def _translate(context, text, disambig):
        return QtGui.QApplication.translate(context, text, disambig, _encoding)
except AttributeError:
    def _translate(context, text, disambig):
        return QtGui.QApplication.translate(context, text, disambig)

class Ui_Settings(object):
    def setupUi(self, Settings):
        Settings.setObjectName(_fromUtf8("Settings"))
        Settings.resize(400, 194)
        self.gridLayout = QtGui.QGridLayout(Settings)
        self.gridLayout.setObjectName(_fromUtf8("gridLayout"))
        self.buttonBox = QtGui.QDialogButtonBox(Settings)
        self.buttonBox.setOrientation(QtCore.Qt.Horizontal)
        self.buttonBox.setStandardButtons(QtGui.QDialogButtonBox.Cancel|QtGui.QDialogButtonBox.Ok)
        self.buttonBox.setObjectName(_fromUtf8("buttonBox"))
        self.gridLayout.addWidget(self.buttonBox, 1, 0, 1, 1)
        self.gridLayout_2 = QtGui.QGridLayout()
        self.gridLayout_2.setObjectName(_fromUtf8("gridLayout_2"))
        self.groupBox = QtGui.QGroupBox(Settings)
        self.groupBox.setObjectName(_fromUtf8("groupBox"))
        self.gridLayout_3 = QtGui.QGridLayout(self.groupBox)
        self.gridLayout_3.setObjectName(_fromUtf8("gridLayout_3"))
        self.lesscPath = QtGui.QLineEdit(self.groupBox)
        self.lesscPath.setObjectName(_fromUtf8("lesscPath"))
        self.gridLayout_3.addWidget(self.lesscPath, 0, 0, 1, 1)
        self.chooseLessc = QtGui.QPushButton(self.groupBox)
        self.chooseLessc.setObjectName(_fromUtf8("chooseLessc"))
        self.gridLayout_3.addWidget(self.chooseLessc, 0, 1, 1, 1)
        self.gridLayout_2.addWidget(self.groupBox, 0, 0, 1, 1)
        self.groupBox_2 = QtGui.QGroupBox(Settings)
        self.groupBox_2.setObjectName(_fromUtf8("groupBox_2"))
        self.gridLayout_4 = QtGui.QGridLayout(self.groupBox_2)
        self.gridLayout_4.setObjectName(_fromUtf8("gridLayout_4"))
        self.label = QtGui.QLabel(self.groupBox_2)
        self.label.setObjectName(_fromUtf8("label"))
        self.gridLayout_4.addWidget(self.label, 0, 0, 1, 1)
        self.editor = QtGui.QLineEdit(self.groupBox_2)
        self.editor.setObjectName(_fromUtf8("editor"))
        self.gridLayout_4.addWidget(self.editor, 0, 1, 1, 1)
        self.chooseEditor = QtGui.QPushButton(self.groupBox_2)
        self.chooseEditor.setObjectName(_fromUtf8("chooseEditor"))
        self.gridLayout_4.addWidget(self.chooseEditor, 0, 2, 1, 1)
        self.lessFolder = QtGui.QCheckBox(self.groupBox_2)
        self.lessFolder.setObjectName(_fromUtf8("lessFolder"))
        self.gridLayout_4.addWidget(self.lessFolder, 1, 1, 1, 1)
        self.gridLayout_2.addWidget(self.groupBox_2, 1, 0, 1, 1)
        self.gridLayout.addLayout(self.gridLayout_2, 0, 0, 1, 1)

        self.retranslateUi(Settings)
        QtCore.QObject.connect(self.buttonBox, QtCore.SIGNAL(_fromUtf8("accepted()")), Settings.accept)
        QtCore.QObject.connect(self.buttonBox, QtCore.SIGNAL(_fromUtf8("rejected()")), Settings.reject)
        QtCore.QMetaObject.connectSlotsByName(Settings)

    def retranslateUi(self, Settings):
        Settings.setWindowTitle(_translate("Settings", "Plessc Setting", None))
        self.groupBox.setTitle(_translate("Settings", "Lessc Path", None))
        self.chooseLessc.setText(_translate("Settings", "Choose", None))
        self.groupBox_2.setTitle(_translate("Settings", "Editor config", None))
        self.label.setText(_translate("Settings", "Editor", None))
        self.chooseEditor.setText(_translate("Settings", "Choose", None))
        self.lessFolder.setText(_translate("Settings", "Open all less files in the folder", None))



ui_MainWindow.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file '/home/mte90/Desktop/Prog/Plessc/MainWindow.ui'
#
# Created: Sat Mar  1 01:11:59 2014
#      by: PyQt4 UI code generator 4.10.3
#
# WARNING! All changes made in this file will be lost!

from PyQt4 import QtCore, QtGui

try:
    _fromUtf8 = QtCore.QString.fromUtf8
except AttributeError:
    def _fromUtf8(s):
        return s

try:
    _encoding = QtGui.QApplication.UnicodeUTF8
    def _translate(context, text, disambig):
        return QtGui.QApplication.translate(context, text, disambig, _encoding)
except AttributeError:
    def _translate(context, text, disambig):
        return QtGui.QApplication.translate(context, text, disambig)

class Ui_MainWindow(object):
    def setupUi(self, MainWindow):
        MainWindow.setObjectName(_fromUtf8("MainWindow"))
        MainWindow.resize(531, 459)
        icon = QtGui.QIcon()
        icon.addPixmap(QtGui.QPixmap(_fromUtf8("logo.png")), QtGui.QIcon.Normal, QtGui.QIcon.Off)
        MainWindow.setWindowIcon(icon)
        self.centralwidget = QtGui.QWidget(MainWindow)
        self.centralwidget.setObjectName(_fromUtf8("centralwidget"))
        self.gridLayout_2 = QtGui.QGridLayout(self.centralwidget)
        self.gridLayout_2.setObjectName(_fromUtf8("gridLayout_2"))
        self.gridLayout = QtGui.QGridLayout()
        self.gridLayout.setObjectName(_fromUtf8("gridLayout"))
        self.horizontalLayout = QtGui.QHBoxLayout()
        self.horizontalLayout.setObjectName(_fromUtf8("horizontalLayout"))
        self.info = QtGui.QLabel(self.centralwidget)
        self.info.setMinimumSize(QtCore.QSize(400, 0))
        self.info.setText(_fromUtf8(""))
        self.info.setObjectName(_fromUtf8("info"))
        self.horizontalLayout.addWidget(self.info)
        self.compile = QtGui.QPushButton(self.centralwidget)
        self.compile.setObjectName(_fromUtf8("compile"))
        self.horizontalLayout.addWidget(self.compile)
        self.gridLayout.addLayout(self.horizontalLayout, 3, 0, 1, 1)
        self.horizontalLayout_2 = QtGui.QHBoxLayout()
        self.horizontalLayout_2.setObjectName(_fromUtf8("horizontalLayout_2"))
        self.verticalGroupBox = QtGui.QGroupBox(self.centralwidget)
        self.verticalGroupBox.setObjectName(_fromUtf8("verticalGroupBox"))
        self.verticalLayout = QtGui.QVBoxLayout(self.verticalGroupBox)
        self.verticalLayout.setObjectName(_fromUtf8("verticalLayout"))
        self.setMinify = QtGui.QCheckBox(self.verticalGroupBox)
        self.setMinify.setObjectName(_fromUtf8("setMinify"))
        self.verticalLayout.addWidget(self.setMinify)
        self.line = QtGui.QFrame(self.verticalGroupBox)
        self.line.setFrameShape(QtGui.QFrame.HLine)
        self.line.setFrameShadow(QtGui.QFrame.Sunken)
        self.line.setObjectName(_fromUtf8("line"))
        self.verticalLayout.addWidget(self.line)
        self.optionIE = QtGui.QCheckBox(self.verticalGroupBox)
        self.optionIE.setObjectName(_fromUtf8("optionIE"))
        self.verticalLayout.addWidget(self.optionIE)
        self.label_3 = QtGui.QLabel(self.verticalGroupBox)
        self.label_3.setStyleSheet(_fromUtf8(""))
        self.label_3.setTextFormat(QtCore.Qt.RichText)
        self.label_3.setOpenExternalLinks(True)
        self.label_3.setObjectName(_fromUtf8("label_3"))
        self.verticalLayout.addWidget(self.label_3)
        self.line_2 = QtGui.QFrame(self.verticalGroupBox)
        self.line_2.setFrameShape(QtGui.QFrame.HLine)
        self.line_2.setFrameShadow(QtGui.QFrame.Sunken)
        self.line_2.setObjectName(_fromUtf8("line_2"))
        self.verticalLayout.addWidget(self.line_2)
        self.optionSourceMap = QtGui.QCheckBox(self.verticalGroupBox)
        self.optionSourceMap.setObjectName(_fromUtf8("optionSourceMap"))
        self.verticalLayout.addWidget(self.optionSourceMap)
        self.horizontalLayout_2.addWidget(self.verticalGroupBox)
        self.verticalGroupBox1 = QtGui.QGroupBox(self.centralwidget)
        self.verticalGroupBox1.setObjectName(_fromUtf8("verticalGroupBox1"))
        self.verticalLayout_2 = QtGui.QVBoxLayout(self.verticalGroupBox1)
        self.verticalLayout_2.setObjectName(_fromUtf8("verticalLayout_2"))
        self.setStandard = QtGui.QRadioButton(self.verticalGroupBox1)
        self.setStandard.setChecked(True)
        self.setStandard.setObjectName(_fromUtf8("setStandard"))
        self.verticalLayout_2.addWidget(self.setStandard)
        self.setBoth = QtGui.QRadioButton(self.verticalGroupBox1)
        self.setBoth.setObjectName(_fromUtf8("setBoth"))
        self.verticalLayout_2.addWidget(self.setBoth)
        self.line_3 = QtGui.QFrame(self.verticalGroupBox1)
        self.line_3.setFrameShape(QtGui.QFrame.HLine)
        self.line_3.setFrameShadow(QtGui.QFrame.Sunken)
        self.line_3.setObjectName(_fromUtf8("line_3"))
        self.verticalLayout_2.addWidget(self.line_3)
        self.lint = QtGui.QPushButton(self.verticalGroupBox1)
        self.lint.setObjectName(_fromUtf8("lint"))
        self.verticalLayout_2.addWidget(self.lint)
        self.horizontalLayout_2.addWidget(self.verticalGroupBox1)
        self.gridLayout.addLayout(self.horizontalLayout_2, 1, 0, 1, 1)
        self.gridLayout_3 = QtGui.QGridLayout()
        self.gridLayout_3.setObjectName(_fromUtf8("gridLayout_3"))
        self.inputEdit = QtGui.QPushButton(self.centralwidget)
        self.inputEdit.setObjectName(_fromUtf8("inputEdit"))
        self.gridLayout_3.addWidget(self.inputEdit, 0, 3, 1, 1)
        self.inputChoose = QtGui.QPushButton(self.centralwidget)
        self.inputChoose.setObjectName(_fromUtf8("inputChoose"))
        self.gridLayout_3.addWidget(self.inputChoose, 0, 2, 1, 1)
        self.label = QtGui.QLabel(self.centralwidget)
        self.label.setObjectName(_fromUtf8("label"))
        self.gridLayout_3.addWidget(self.label, 0, 0, 1, 1)
        self.label_2 = QtGui.QLabel(self.centralwidget)
        self.label_2.setObjectName(_fromUtf8("label_2"))
        self.gridLayout_3.addWidget(self.label_2, 1, 0, 1, 1)
        self.outputFile = QtGui.QLineEdit(self.centralwidget)
        self.outputFile.setObjectName(_fromUtf8("outputFile"))
        self.gridLayout_3.addWidget(self.outputFile, 1, 1, 1, 1)
        self.outputChoose = QtGui.QPushButton(self.centralwidget)
        self.outputChoose.setObjectName(_fromUtf8("outputChoose"))
        self.gridLayout_3.addWidget(self.outputChoose, 1, 2, 1, 1)
        self.outputLog = QtGui.QPushButton(self.centralwidget)
        self.outputLog.setObjectName(_fromUtf8("outputLog"))
        self.gridLayout_3.addWidget(self.outputLog, 1, 3, 1, 1)
        self.inputFile = QtGui.QLineEdit(self.centralwidget)
        self.inputFile.setObjectName(_fromUtf8("inputFile"))
        self.gridLayout_3.addWidget(self.inputFile, 0, 1, 1, 1)
        self.gridLayout.addLayout(self.gridLayout_3, 0, 0, 1, 1)
        self.autoCompile = QtGui.QCheckBox(self.centralwidget)
        self.autoCompile.setObjectName(_fromUtf8("autoCompile"))
        self.gridLayout.addWidget(self.autoCompile, 2, 0, 1, 1)
        self.log = QtWebKit.QWebView(self.centralwidget)
        self.log.setUrl(QtCore.QUrl(_fromUtf8("about:blank")))
        self.log.setObjectName(_fromUtf8("log"))
        self.gridLayout.addWidget(self.log, 4, 0, 1, 1)
        self.gridLayout_2.addLayout(self.gridLayout, 0, 0, 1, 1)
        MainWindow.setCentralWidget(self.centralwidget)
        self.menubar = QtGui.QMenuBar(MainWindow)
        self.menubar.setGeometry(QtCore.QRect(0, 0, 531, 22))
        self.menubar.setObjectName(_fromUtf8("menubar"))
        self.menuSetting = QtGui.QMenu(self.menubar)
        self.menuSetting.setObjectName(_fromUtf8("menuSetting"))
        self.menuInfo = QtGui.QMenu(self.menubar)
        self.menuInfo.setObjectName(_fromUtf8("menuInfo"))
        MainWindow.setMenuBar(self.menubar)
        self.actionAbout_Plessc = QtGui.QAction(MainWindow)
        self.actionAbout_Plessc.setObjectName(_fromUtf8("actionAbout_Plessc"))
        self.actionSettings = QtGui.QAction(MainWindow)
        self.actionSettings.setObjectName(_fromUtf8("actionSettings"))
        self.menuSetting.addAction(self.actionSettings)
        self.menuInfo.addAction(self.actionAbout_Plessc)
        self.menubar.addAction(self.menuSetting.menuAction())
        self.menubar.addAction(self.menuInfo.menuAction())

        self.retranslateUi(MainWindow)
        QtCore.QMetaObject.connectSlotsByName(MainWindow)

    def retranslateUi(self, MainWindow):
        MainWindow.setWindowTitle(_translate("MainWindow", "MainWindow", None))
        self.compile.setText(_translate("MainWindow", "Compile it!", None))
        self.verticalGroupBox.setTitle(_translate("MainWindow", "Option ", None))
        self.setMinify.setText(_translate("MainWindow", "Minify", None))
        self.optionIE.setText(_translate("MainWindow", "Disable IE compatibility checks", None))
        self.label_3.setText(_translate("MainWindow", "<html><head/><body><p><a href=\"https://github.com/less/less.js/pull/1190\"><span style=\" text-decoration: underline; color:#0057ae;\">More info for --ie-no-compat parameter</span></a></p></body></html>", None))
        self.optionSourceMap.setText(_translate("MainWindow", "SourceMap", None))
        self.verticalGroupBox1.setTitle(_translate("MainWindow", "Save Method", None))
        self.setStandard.setText(_translate("MainWindow", "Only minified version", None))
        self.setBoth.setText(_translate("MainWindow", "Standard and minified version", None))
        self.lint.setText(_translate("MainWindow", "Lint (Report only errors)", None))
        self.inputEdit.setText(_translate("MainWindow", "Edit", None))
        self.inputChoose.setText(_translate("MainWindow", "Choose", None))
        self.label.setText(_translate("MainWindow", "Input File", None))
        self.label_2.setText(_translate("MainWindow", "Output File", None))
        self.outputChoose.setText(_translate("MainWindow", "Choose", None))
        self.outputLog.setText(_translate("MainWindow", "Open Log", None))
        self.autoCompile.setText(_translate("MainWindow", "Auto Compile - Check if the input less file is updated and compile", None))
        self.menuSetting.setTitle(_translate("MainWindow", "Setting", None))
        self.menuInfo.setTitle(_translate("MainWindow", "Info", None))
        self.actionAbout_Plessc.setText(_translate("MainWindow", "About Plessc", None))
        self.actionSettings.setText(_translate("MainWindow", "Settings", None))

from PyQt4 import QtWebKit


plessc.py
#!/usr/bin/python3
from PyQt4.QtCore import *
from PyQt4.QtGui import *
import os, sys, signal, datetime

from ui_MainWindow import Ui_MainWindow
from settings import SettingDialog

class MainWindow ( QMainWindow , Ui_MainWindow):

    #var initialization
    settings = QSettings('Mte90','Plessc')
    settings.setFallbacksEnabled(False)
    history = QSettings('Mte90','Plessc_History')
    history.setFallbacksEnabled(False)
    version = 'v 1.1'
    input_less = ''
    output_css = ''
    mysize = ''
    history_field = {}
    history_field['input'] = []
    history_field['output'] = []
    option = {}
    proc = QProcess()
    less_version = QProcess()
    watcher = QFileSystemWatcher()

    def __init__ ( self, parent = None ):
        QMainWindow.__init__( self, parent )
        self.ui = Ui_MainWindow()
        self.ui.setupUi( self )
        self.setWindowTitle('PLessc - lessc not defined')
        #Connect the function with the signal
        self.ui.inputChoose.clicked.connect(self.openInputDialog)
        self.ui.outputChoose.clicked.connect(self.openOutputDialog)
        self.ui.setBoth.pressed.connect(self.setBoth)
        self.ui.setStandard.pressed.connect(self.setStandard)
        self.ui.optionIE.stateChanged.connect(self.setOptionIE)
        self.ui.optionSourceMap.stateChanged.connect(self.setOptionSourceMap)
        self.ui.setMinify.stateChanged.connect(self.setMinify)
        self.ui.inputEdit.pressed.connect(self.openEditor)
        self.ui.outputLog.clicked.connect(self.openLog)
        self.ui.lint.clicked.connect(self.lintLog)
        self.ui.compile.clicked.connect(self.compileIt)
        self.ui.menuInfo.triggered.connect(self.openInfo)
        self.ui.menuSetting.triggered.connect(self.openSetDialog)
        self.proc.finished.connect(self.checkLog)
        self.less_version.finished.connect(self.updateTitle)
        signal.signal(signal.SIGINT, signal.SIG_DFL)
        #hide log
        self.ui.log.hide()
        #check lessc version
        self.loadVersion()
        #Check the setting for load the path of the file
        if self.settings.value('input_file') != -1:
            self.input_less = self.settings.value('input_file')
        else:
            self.input_less = '/'
        if self.settings.value('output_file') != -1:
            self.output_css = self.settings.value('output_file')
        else:
            self.output_css = '/'
        self.ui.inputFile.setText(self.input_less)
        self.ui.outputFile.setText(self.output_css)
        #Check the setting for the minify mode for the css file
        if self.settings.value('min') == 'False':
            self.option['minify'] = ' ';
            self.settings.setValue('min','False')
            self.ui.setMinify.setChecked(False)
        else:
            self.option['minify'] = '-x'
            self.settings.setValue('min','True')
            self.ui.setMinify.setChecked(True)
        #Check for the export of the file
        if self.settings.value('both_or_standard') == 'False':
            self.settings.setValue('both_or_standard','False')
            self.ui.setStandard.toggle()
        else:
            self.settings.setValue('both_or_standard','True')
            self.ui.setBoth.toggle()
        #Auto compile option
        if self.settings.value('auto_compile') == 'False':
            self.ui.autoCompile.setChecked(False)
        else:
            self.ui.autoCompile.setChecked(True)
        #Add the function for the autocompile after the load of the setting
        self.ui.autoCompile.stateChanged.connect(self.autoCompile)
        self.autoCompile()
        #Setting for export the css for old IE version
        if self.settings.value('option_IE') == 'False':
            self.ui.optionIE.setChecked(False)
            self.option['ie'] = ' '
        else:
            self.ui.optionIE.setChecked(True)
            self.option['ie'] = '--no-ie-compat'
        #Setting for enable Sourcemap
        if self.settings.value('option_SourceMap') == 'False':
            self.ui.optionSourceMap.setChecked(False)
            self.option['sourcemap'] = ' '
        else:
            self.ui.optionSourceMap.setChecked(True)
            self.option['sourcemap'] = '--source-map'
        #Resize the window for hide the space of log
        self.resize(505,220)
        #Load History
        self.loadHistory()
        self.show()
        
    def loadHistory(self):
        #load history
        if str(self.history.value('input')) != 'None':
            self.history_field['input'] = str(self.history.value('input')).split(';')
        if str(self.history.value('output')) != 'None':
            self.history_field['output'] = str(self.history.value('output')).split(';')
        completer_input = QCompleter(self.history_field['input'],self.ui.inputFile)
        self.ui.inputFile.setCompleter(completer_input)
        completer_output = QCompleter(self.history_field['output'],self.ui.outputFile)
        self.ui.outputFile.setCompleter(completer_output)
        self.ui.inputFile.editingFinished.connect(self.setInputFile)
        self.ui.outputFile.editingFinished.connect(self.setOutputFile)

    def addHistory(self,field,text):
        #add the path in the history if not exist
        if text not in self.history_field[field]:
            self.history_field[field].append(text)
            self.history.setValue(field,';'.join(self.history_field[field]))

    #Function for open a dialog for choose the input less file
    def openInputDialog(self):
        self.input_less = QFileDialog.getOpenFileName(self, 'Choose less file',self.ui.inputFile.text(),'LESS file (*.less)')
        if self.input_less != '':
             self.ui.inputFile.setText(self.input_less)
    
    #Function for open a dialog for choose the output css file
    def openOutputDialog(self):
        self.output_css = QFileDialog.getSaveFileName(self, 'Set css file',self.ui.outputFile.text(),'CSS file (*.css)')
        if self.output_css != '':
             self.ui.outputFile.setText(self.output_css)
    
    #Save the input file in the setting
    def setInputFile(self):
        self.settings.setValue('input_file',self.ui.inputFile.text())
        self.addHistory('input',self.ui.inputFile.text())
    
    #Save the output file in the setting
    def setOutputFile(self):
        self.settings.setValue('output_file',self.ui.outputFile.text())
        self.addHistory('output',self.ui.outputFile.text())
    
    #Save the output export of the css
    def setBoth(self):
        self.settings.setValue('both_or_standard','True')
    
    #Save the output export of the css
    def setStandard(self):
        self.settings.setValue('both_or_standard','False')
    
    #Check autoCompile and enable the watching of the input less file
    def autoCompile(self):
        #If not checked
        if self.ui.autoCompile.isChecked() == False:
            self.settings.setValue('auto_compile','False')
            #If previosuly was enabled the option this disable the watch of the file
            try:
                self.watcher.fileChanged.disconnect()
            except (RuntimeError, TypeError, NameError):
                pass
        else:
            self.settings.setValue('auto_compile','True')
            #Clean the path added to file watching for fix a problem with Qt4
            self.watcher.removePath(self.settings.value('input_file'))
            #Re add of the path
            self.watcher.addPath(self.settings.value('input_file'))
            self.watcher.fileChanged.connect(self.compileIt)
    
    #Save the Ie Setting
    def setOptionIE(self):
        if self.ui.optionIE.isChecked() == False:
            self.settings.setValue('option_IE','False')
            self.option['ie'] = ''
        else:
            self.settings.setValue('option_IE','True')
            self.option['ie'] = '--no-ie-compat'
    
    #Save the SourceMap
    def setOptionSourceMap(self):
        if self.ui.optionSourceMap.isChecked() == False:
            self.settings.setValue('option_SourceMap','False')
            self.option['sourcemap'] = ' '
        else:
            self.settings.setValue('option_SourceMap','True')
            self.option['sourcemap'] = '--source-map'
            
    #Save the minify setting
    def setMinify(self):
        if self.ui.setMinify.isChecked() == False:
            self.option['minify'] = ' '
            self.settings.setValue('min','False')
        else:
            self.option['minify'] = '-x'
            self.settings.setValue('min','True')
    
    #Compile the less file
    def compileIt(self):
        QApplication.setOverrideCursor(QCursor(Qt.WaitCursor))
        if os.path.isfile(self.settings.value('input_file')):
            self.ui.log.setHtml('')
            self.ui.info.setText('Compiling...')
            if self.settings.value('both_or_standard') == 'True':
                #if both save method True
                name = os.path.splitext(self.settings.value('output_file'))[0]
                name += '.min.css'
                complete = str(self.settings.value('less_path') + self.optionString() + '"' + self.settings.value('input_file') + '" "' + name + '"' )
                command = str(self.settings.value('less_path') + self.optionString('minify') + ' --verbose "' + self.settings.value('input_file') + '" "' + self.settings.value('output_file') + '"')
                #Compile the min.css
                os.system(complete)
                self.proc.closeWriteChannel()
                #Compile a standard css
                self.proc.start(command)
                self.proc.waitForFinished()
                self.proc.closeWriteChannel()
                if os.path.isfile(name):
                    self.ui.info.setText('File Min Output: <b>' + self.sizeof_fmt(name) + '</b> | File Standard: <b>' + self.sizeof_fmt(self.settings.value('output_file')) + '</b> | ' + datetime.datetime.now().strftime("%d-%m-%Y %H:%M:%S"))
                else:
                    QMessageBox.critical(self.window(), "Output File not exist","The output file choosen not exist!")
            else:
                #if standard = 0 False
                command = str(self.settings.value('less_path') + self.optionString() + ' --verbose "' + self.settings.value('input_file') + '" "' + self.settings.value('output_file') + '"' )
                self.proc.closeWriteChannel()
                self.proc.start(command)
                self.proc.waitForFinished()
                self.proc.closeWriteChannel()
                if os.path.isfile(self.settings.value('output_file')):
                    self.ui.info.setText('File Output: <b>' + self.sizeof_fmt(self.settings.value('output_file')) + '</b>' + ' | ' + datetime.datetime.now().strftime("%d-%m-%Y %H:%M:%S"))
                else:
                    QMessageBox.critical(self.window(), "Output File not exist","The output file choosen not exist!")
            QApplication.restoreOverrideCursor()
        else:
            QMessageBox.critical(self.window(), "Input File not exist","The input file choosen not exist!")
        print(command)
        #Clean the path added to file watching for fix a problem with Qt4
        self.watcher.removePath(self.settings.value('input_file'))
        #Readd the path
        self.watcher.addPath(self.settings.value('input_file'))
    
    #Open all the less file in the folder onf the input file
    def openEditor(self):
        open_file = self.settings.value('input_file')
        #get all file less and open on the editor if this option are set
        if self.settings.value('less_folder') == 'True':
            list_file = ''
            path_less = os.path.split(str(open_file))[0]
            for root, dirs, files in os.walk(path_less):
                files.sort()
                for name in files:
                    filename = os.path.join(root, name)
                    if filename.endswith('.less'):
                        list_file = list_file + '"' + filename + '" '
            open_file = list_file
        os.system(str(self.settings.value('editor_path') + ' ' + open_file))
    
    #Show the log
    def openLog(self):
        if self.ui.log.isVisible() == True:
            self.resize(531,239)
            self.ui.log.hide()
        else:
            self.resize(531,459)
            self.ui.log.show()

    def openSetDialog(self):
        #i need to explain this??
        window = QDialog()
        ui = SettingDialog()
        ui.setupUi(window)
        if ui.exec_() == 1:
            self.loadVersion()

    def openInfo(self):
        QMessageBox.about(self.window(), "About Plessc","<p align='center'>Plessc " + self.version + " <br><br>By <a href='http://www.mte90.net'><b>Mte90</b></a><br><br>GUI in Python and Qt4 for compile less file<br><br>Tested with lessc 1.7.x of LESS.JS<br><br><small>If other compilers use the same parameters i think that works else some monkeys do it for you</small><br><br>License: GPL 3</p>")

    #http://stackoverflow.com/questions/1094841/reusable-library-to-get-human-readable-version-of-file-size
    def sizeof_fmt(self,file_):
        num = os.path.getsize(str(file_))
        for x in ['bytes','KB','MB','GB','TB']:
            if num < 1024.0:
                return str("%3.1f %s") % (num, x)
            num /= 1024.0
    
    #Clean the output of lessc 
    def replace_all(self,text):
        #remove the shellcode/ANSI of the color of the text
        text = text.replace('[39m', '<br>').replace('[31m', '').replace('[22m', '').replace('[0m', '').replace('1b', '')
        text = text.replace('[90m', '').replace('[27m', '').replace('[7m', '').replace('[1m', '').replace("b''",'').replace('\n\n','').replace('b\'','')
        text = text.replace('\\x', '').replace('\\n\\n\'', '').replace('\\n', '')
        
        return text.lstrip()
        
    #Execeute the lint and show it in the log area
    def lintLog(self):
        self.proc.closeWriteChannel()
        QApplication.setOverrideCursor(QCursor(Qt.WaitCursor))
        self.proc.start(str(self.settings.value('less_path') + ' --lint "' + self.settings.value('input_file') + '"'))
        self.openLog()
        self.proc.closeWriteChannel()
        QApplication.restoreOverrideCursor()
    
    #Check the content of log
    def checkLog(self):
        stdout = str(self.proc.readAllStandardOutput())
        check = stdout.strip()
        #If the log not exmpty and length > of 3 print
        if(check is not None and len(check) > 3):
            output = self.replace_all(stdout)
            if not output.startswith('lessc: wrote '):
                self.ui.log.setHtml(self.replace_all(stdout))
                self.openLog()
        else:
            self.ui.log.setHtml('OK!')
    
    #Execute the load of lessc version
    def loadVersion(self):
        self.less_version.closeWriteChannel()
        self.setWindowTitle('PLessc - lessc not defined')
        self.less_version.start(self.settings.value('less_path'),['--version'])
        
    #Update the title with less version
    def updateTitle(self):
        stdout = str(self.less_version.readAllStandardOutput())
        self.setWindowTitle('PLessc - ' + self.replace_all(stdout.rstrip('\'')))
    
    #Concate all the settings for lessc for use it in the command
    def optionString(self,remove=[]):
        #remove the parameter
        if remove != []:
            del self.option[remove]
        string = ' '.join('{}'.format(val) for key, val in self.option.items())
        return ' ' + string + ' '
        
def main():
    app = QApplication(sys.argv)
    MainWindow_ = QMainWindow()
    ui = MainWindow()
    ui.setupUi(MainWindow_)
    sys.exit(app.exec_())

main()